keps/sig-node/4603-tune-crashloopbackoff/README.md

```diff
@@ -0,0 +1,1222 @@
+
+
+
+**Alternative**: Workloads must opt-in with `restartPolicy: FastOnSuccess`, as a
+foil to `restartPolicy: OnFailure`. In this case, existing workloads with
+`restartPolicy: Always` or ones not determined to be in the critical path would
+use the new, yet still relatively slower, front-loaded decay curve and only
+those updated with `FastOnSuccess` would get truer fast restart behavior.
+However, then it becomes impossible for a workload to opt into both
+`restartPolicy: FastOnSuccess` and `restartPolicy: Rapid`.
+
+#### API opt in for max cap decay curve (`restartPolicy: Rapid`)
+
+For some users in
+[Kubernetes#57291](https://github.com/kubernetes/kubernetes/issues/57291), any
+delay over 1 minute at any point is just too slow, even if it is legitimately
+crashing. A common refrain is that for independently recoverable errors,
+especially system infrastructure events or recovered external dependencies, or
+for absolutely nonnegotiably critical sidecar pods, users would rather poll more
+often or more intelligently to reduce the amount of time a workload has to wait
+to try again after a failure. In the extreme cases, users want to be able to
+configure (by container, node, or exit code) the backoff to close to 0 seconds.
+This KEP considers it out of scope to implement fully user-customizable
+behavior, and too risky to node stability to allow legitimately crashing
+workloads to have a backoff of 0, but it is in scope to provide users a way to
+opt workloads in to a faster restart curve that is not as drastic as what is
+intended for `Success` states, nor as beholden to the status quo as the new
+default front loaded decay with interval modification.
+
+Pods and restartable init (aka sidecar) containers will be able to set a new
+OneOf value, `restartPolicy: Rapid`, to opt in to an exponential backoff decay
+that starts at a low initial value and maximizes to a cap of 1 minute. The
+detailed methodology for determining the implementable starting value, and
+benchmarking it during and after alpha, is enclosed in Design Details, but will
+start at 250ms.
+
+TODO: Confirm the interaction between pod level restart policy and container
+level restartpolicy if not only for sidecar containers
+
+
+#### Observability
+
+Again, let it be known that by definition this KEP will cause pods to restart
+faster and more often than the current status quo and such a change is desired.
+However, to do so as safely as possible, improved visibility into cluster
+restart behavior is needed both for benchmarking this change and for cluster
+operators to be able to quantify the risk posed to their specific clusters on
+upgrade.
+
+This KEP requires the ability to determine, for a given percentage of
+heterogenity between "Succeeded" terminating pods, crashing pods whose
+`restartPolicy: Always`, and crashing pods whose `restartPolicy: Rapid`, 
+ * what is the load and rate of Pod restart related API requests to the API
+   server?
```

Comment:
This should be deterministic, right? As in, we should be able to step through the code and see the exact number of API requests the kubelet makes in the course of a pod restart, right? That seems like valuable data, but not sure we need observability for it?

Reply:
You're right, I updated this some, especially in light of other changes/back of the napkin calculations to Risks and Mitigations, in [8caa42a](https://github.com/kubernetes/enhancements/pull/4604/commits/8caa42a4518dfb105817bf00065d517b5dd4ab5b), to be more about the overall benchmarking plan.